[FSDP2] Added test to show rank 0 broadcast for HSDP replicas

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #125484
* #125479
* __->__ #125431

This PR shows a simple utility to broadcast the parameters across replicas for HSDP:
```
replicate_group = mesh.get_group("replicate")
for param in model.parameters():
    # E.g. for mesh [[0, 1, 2, 3], [4, 5, 6, 7]] sharding on dim-1 and
    # replicating on dim-0, broadcast with sources 0, 1, 2, 3
    src_rank = dist.get_process_group_ranks(replicate_group)[0]
    torch.distributed.broadcast(
        param.to_local(), src=src_rank, group=replicate_group
    )
```


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

Commit:
[FSDP2] Added test to show rank 0 broadcast for HSDP replicas

[ghstack-poisoned]

awgu commented on test/distributed/_composable/fsdp/test_fully_shard_init.py:
```diff
@@ -725,5 +725,69 @@ def test_process_group_init(self):
+            # E.g. for mesh [[0, 1, 2, 3], [4, 5, 6, 7]] sharding on dim-1 and
+            # replicating on dim-0, broadcast with sources 0, 1, 2, 3
+            src_rank = dist.get_process_group_ranks(replicate_group)[0]
+            torch.distributed.broadcast(
```
Today, in-place c10d broadcast is preferred.

If we want to use functional broadcast:
1. We need to verify the semantics. We may still need to get the `src_rank` like we do here, which is confusing since it is the rank with respect to the global process group, not the one passed to broadcast.
2. We need to swap the newly broadcasted tensor in. Since FSDP holds a reference, we cannot just `setattr(module, param_name, broadcasted_param)` since that would leave FSDP's reference as stale. We may consider using `swap_tensors`, but we are blocked by the local tensor padding issue since the broadcasted parameter would not have padding and is actually a view into the padded local tensor.

wanchaol replied:
Yeah I think a inplace broadcast make sense here!

awgu commented on test/distributed/_composable/fsdp/test_fully_shard_init.py:
```diff
@@ -725,5 +725,69 @@ def test_process_group_init(self):
+        return 4
+
+    @unittest.skipIf(not TEST_CUDA, "no cuda")
+    def test_hsdp_broadcast_across_replicas(self):
```
We might wonder, why not _always_ have HSDP broadcast during init time. The issue is that we only need to broadcast if we are initializing from scratch (not a checkpoint). If we are initializing from a checkpoint, then we are already guaranteed that replicas are the same, and broadcasting is wasteful and can affect init time.

weifengpy commented on test/distributed/_composable/fsdp/test_fully_shard_init.py:
```diff
@@ -725,5 +725,69 @@ def test_process_group_init(self):
+
+        # Only preserve the model parameters on the replicate mesh's rank 0
+        if mesh.get_local_rank("replicate") > 0:
+            print(f"[Rank {self.rank}] filling with 1337")
```
is the `print` meant to land?

weifengpy reviewed (approved):
nice! I feel good to have the exmaple of getting global rank from `replicate_group`. hopefully people copy from here instead of using local rank in c10d

Commit:
Update on "[FSDP2] Added test to show rank 0 broadcast for HSDP replicas"


This PR shows a simple utility to broadcast the parameters across replicas for HSDP:
```
replicate_group = mesh.get_group("replicate")
for param in model.parameters():
    # E.g. for mesh [[0, 1, 2, 3], [4, 5, 6, 7]] sharding on dim-1 and
    # replicating on dim-0, broadcast with sources 0, 1, 2, 3
    src_rank = dist.get_process_group_ranks(replicate_group)[0]
    torch.distributed.broadcast(
        param.to_local(), src=src_rank, group=replicate_group
    )
```


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

awgu commented:
By the way, I am open to the idea of some `fsdp/utils.py` file for some common code snippets like this. However, in my personal opinion, I think that (1) we should not do this by default in FSDP and (2) we should not make it a flag.

(1) is because it is a waste to broadcast if loading from a state dict.
(2) is because it makes this inflexible in the user-code, as the user must know at FSDP sharding time whether it will load a state dict or not. We can imagine the code to be cleaner if we have the shared FSDP sharding code and then later when choosing to load from state dict or init weights, we decide whether to broadcast or not.

weifengpy commented:
> (1) we should not do this by default in FSDP and (2) we should not make it a flag.

(1) feels easy to understand
(2) 'make it a flag' means a ENV flag or 'make it an official best practice'?

awgu commented:
> (2) 'make it a flag' means a ENV flag or 'make it an official best practice'?

Today, FSDP1 has a `sync_module_states: bool` flag to control this kind of behavior. I would prefer to not add an analogous flag to FSDP2. We should definitely socialize properly / document that _if_ you are pretraining from scratch with HSDP, you _should_ broadcast across replicas (unless you know your seed is set properly).

wanchaol reviewed (approved):
lgtm

awgu commented:
@pytorchbot merge
